File: monogame_space_invaders-fsharp-master/Player.fs
module Player

open Microsoft.Xna.Framework
open Microsoft.Xna.Framework.Graphics
open Microsoft.Xna.Framework.Input
open Microsoft.Xna.Framework.Content

type Player (pos: Vector2) = class
    
    //private variables
    let mutable _TileBoundingBox: Rectangle = Rectangle(0, 0, 0, 0)
    let mutable _sprite: Texture2D = null
    let mutable keyState = KeyboardState()

    let mutable cbuttonState = GamePadState() 
    let mutable lastcbuttonState = GamePadState()
    let mutable cpadState = GamePadState()
    
    let move_speed: float32 = 350.0f

    //public variables
    member val dead = false with get, set
    member val shoot = false with get, set
    member val position = pos with get, set
    member this.TileBoundingBox = _TileBoundingBox

    //functions    
    member this.loadResources(con: ContentManager) =
        _sprite <- con.Load<Texture2D>("player");
        this.position <- Vector2(640.0f / 2.0f - (float32) _sprite.Bounds.Width / 2.0f, this.position.Y)
        _TileBoundingBox <- Rectangle((int)pos.X, (int) pos.Y, _sprite.Width, _sprite.Height)

    member this.input(gameTime: GameTime) =

        if this.dead = false 
            then

                cpadState <- GamePad.GetState(PlayerIndex.One)         
                lastcbuttonState <-  cbuttonState

                if Keyboard.GetState().IsKeyDown(Keys.Space) && keyState.IsKeyDown(Keys.Space) = false 
                    || GamePad.GetState(PlayerIndex.One).IsButtonDown(Buttons.A) && lastcbuttonState.IsButtonUp(Buttons.A)
                    then
                        this.shoot <- true
                    else
                        this.shoot <- false

                cbuttonState <- cpadState

                if Keyboard.GetState().IsKeyDown(Keys.A) || Keyboard.GetState().IsKeyDown(Keys.Left) || cpadState.ThumbSticks.Left.X <= -0.5f
                    then this.position <- Vector2(this.position.X - move_speed * (float32) gameTime.ElapsedGameTime.TotalSeconds, this.position.Y)

                if Keyboard.GetState().IsKeyDown(Keys.D) || Keyboard.GetState().IsKeyDown(Keys.Right) || cpadState.ThumbSticks.Left.X >= 0.5f
                    then this.position <- Vector2(this.position.X + move_speed * (float32) gameTime.ElapsedGameTime.TotalSeconds, this.position.Y)

                keyState <- Keyboard.GetState()

    member this.update(gameTime: GameTime) =
        this.input(gameTime)

        if this.dead = false 
            then           
            this.position <- Vector2(MathHelper.Clamp(this.position.X, 0.0f, (float32) 640 - (float32) _sprite.Width), 
                                    MathHelper.Clamp(this.position.Y, 0.0f, (float32) 480 - (float32) _sprite.Height));

        _TileBoundingBox <- Rectangle((int)this.position.X, (int) this.position.Y, _sprite.Width, _sprite.Height)

    member this.draw(batch: Microsoft.Xna.Framework.Graphics.SpriteBatch) =
        if this.dead = false
            then batch.Draw(_sprite, this.position, Color.White);
end

File: monogame_space_invaders-desktop-master/Game/Player.cs
using System;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Content;
using Microsoft.Xna.Framework.Input;
using System.Diagnostics;

namespace SpaceInvaders_Desktop
{
    class Player
    {
        Vector2 pos;
        private Texture2D tex;
        const int speed = 250;

        public bool shoot;

        float spriteWidth;

        GamePadState keyState;
        GamePadState lastkeyState;
        GamePadState gamePadState;

        public bool dead;

        public Player()
        {
            dead = false;
        }

        public virtual Rectangle TileBoundingBox
        {
            get
            {
                return new Rectangle(
                    (int)pos.X,
                    (int)pos.Y,
                    28,
                    21);
            }
        }

        public Vector2 Position
        {
            get { return pos; }
            set { pos = value; }
        }

        public float SpriteWidth
        {
            get { return spriteWidth; }
        }

        public void LoadResources(ContentManager con)
        {
            tex = con.Load<Texture2D>("player");
            pos = new Vector2(640 / 2 - tex.Bounds.Width / 2, (480 - 60) - tex.Height / 2);
            spriteWidth = tex.Bounds.Width;
        }

        private void Input(GameTime gameTime)
        {
            gamePadState = GamePad.GetState(PlayerIndex.One);
            lastkeyState = keyState;

            if (GamePad.GetState(PlayerIndex.One).IsButtonDown(Buttons.A) && lastkeyState.IsButtonUp(Buttons.A))
            {
                shoot = true;
            }
            else
            {
                shoot = false;
            }

            keyState = gamePadState;

            if (gamePadState.ThumbSticks.Left.X >= 0.5f)
            {
                pos.X += speed * (float)gameTime.ElapsedGameTime.TotalSeconds;
            }
            else if (gamePadState.ThumbSticks.Left.X <= -0.5f)
            {
                pos.X -= speed * (float)gameTime.ElapsedGameTime.TotalSeconds; 
            }

        }

        public void Update(GameTime gameTime)
        {
            if (!dead)
            {
                Input(gameTime);
                pos = new Vector2(MathHelper.Clamp(pos.X, 0, 640 - tex.Width), pos.Y);
                pos = new Vector2(pos.X, (480 - 60) - tex.Height / 2);
            }
        }

        public void Draw(SpriteBatch sp)
        {
            if(!dead)
            sp.Draw(tex, pos, Color.White);
        }
    }
}
